File: csharp/behavioral/chain_of_responsibility.cs
namespace Gof.ModernCSharp.Behavioral;

// Chain of Responsibility passes a request through handlers until one can answer.
// Modern C# note: common as ASP.NET Core middleware, validators, filters, and
// pipelines. Keep ordering explicit.
public static class ChainOfResponsibility
{
    sealed record Request(string? User, string Role);
    abstract class Handler(Handler? next = null)
    {
        public virtual string? Handle(Request request) => next?.Handle(request);
    }

    sealed class AuthHandler(Handler next) : Handler(next)
    {
        public override string? Handle(Request request) =>
            request.User is null ? "missing user" : base.Handle(request);
    }

    sealed class RoleHandler : Handler
    {
        public override string? Handle(Request request) => request.Role == "admin" ? "approved" : "forbidden";
    }

    public static string Run() => new AuthHandler(new RoleHandler()).Handle(new("Ada", "admin"))!;
}

File: py/behavioral/chain_of_responsibility.py
# Chain of Responsibility passes a request through handlers until one can answer.
# Modern Python note: this is common as middleware, hooks, pipelines, and validators.
# Keep each handler small and make ordering explicit.
class Handler:
    def __init__(self, next_handler=None):
        self.next_handler = next_handler

    def handle(self, request):
        if self.next_handler:
            return self.next_handler.handle(request)
        return None


class AuthHandler(Handler):
    def handle(self, request):
        if not request.get("user"):
            return "missing user"
        return super().handle(request)


class RoleHandler(Handler):
    def handle(self, request):
        if request.get("role") != "admin":
            return "forbidden"
        return "approved"


if __name__ == "__main__":
    chain = AuthHandler(RoleHandler())
    print(chain.handle({"user": "Ada", "role": "admin"}))
